File: start/leetcode/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] forward=new int[nums.length];
        int [] reverse =new int[nums.length];
        for(int i=0;i<nums.length;i++){
            if(i==0){
                forward[i]=1;
                reverse[nums.length-1]=1;
            }
            else{
                forward[i]=forward[i-1]*nums[i-1];
                reverse[nums.length-1-i]=reverse[nums.length-i]*nums[nums.length-i];
            }
        }

        int[] result=new int[nums.length];
        for(int i=0;i<nums.length;i++){
            if(i==0)
                result[i]=reverse[i];
            else if(i==nums.length-1)
                result[i]=forward[i];
            else
                result[i]=forward[i]*reverse[i];
        }
        return result;

    }
}

File: start/leetcode/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        forward  = [0]*len(nums)
        reverse  = [0]*len(nums)
        forward[0] =1
        reverse[-1] =1
        
        for i in range(1,len(nums)):
            forward[i]= nums[i-1]*forward[i-1]
            reverse[len(nums)-1-i]=nums[len(nums)-i]*reverse[len(nums)-i]
            
        ans = [0]*len(nums)
        
        for i in range(len(nums)):
            ans[i] = forward[i]*reverse[i]
            
        return ans
            